### PR TITLE
[[ Bug 20187 ]] Add revScriptDescription property

### PR DESCRIPTION
**Pull request by montegoulding**
This is [this PR](https://github.com/livecode/livecode/pull/5739) rebased on develop-8.1. I will leave both open until we are sure what branch autocomplete needs to be rebased on.

### CODE REVIEW

File: engine/src/exec-interface-object.cpp [hunk 1]
```diff
@@ -4657,7 +4660,11 @@ void MCObject::GetRevAvailableVariables(MCExecContext& ctxt, MCNameRef p_key, MC
     }
     
     // The handler code must be the first char of the string
-    const char_t t_handler_code = MCStringGetNativeCharAtIndex(t_key, 0);
+    char_t t_handler_code = MCStringGetNativeCharAtIndex(t_key, 0);
+    
+    // private handlers are P{F|M}
+    if (t_handler_code == 'P')
+        t_handler_code = MCStringGetNativeCharAtIndex(t_key, 1);
```
Comment (livecodeali) [inline engine/src/exec-interface-object.cpp]: What is the effect of this change?

Comment (montegoulding) [reply engine/src/exec-interface-object.cpp]: Ah, I forgot about that bit... I'll remove from this PR as it's unrelated it was just because the `revAvailableHandlers` returns PF and PM for private handlers. I don't think we will use `revAvailableHandlers` all that much once we have `revScriptDescription` so probably not worth changing.

**Comment by montegoulding** [comment]
@livecodeali I can't see any reason for this to cause a hang in the IDE tests

**Comment by montegoulding** [comment]
Actually... this is hanging on `TestRevEffectiveScriptDescriptionBlockedBehavior` so I must have done something silly

**Comment by montegoulding** [comment]
@livecodeali was just that the test hat `TestSkipIf` which isn't supported. Why that hangs the test instead of writing out an execution failure I'm not sure.

File: engine/src/exec-interface-object.cpp [hunk 2]
```diff
@@ -4704,3 +4707,126 @@ void MCObject::GetRevAvailableVariables(MCExecContext& ctxt, MCNameRef p_key, MC
...
+            {
+                MCAutoValueRef t_description;
+                
+                MCAutoValueRefBase<MCScriptObjectRef> t_script_object_ref;
+                
+                if (t_success)
+                    t_success = MCEngineScriptObjectCreate(t_object_ref -> getobject(), 0, &t_script_object_ref);
+                
+                if (t_success)
+                    t_description = MCEngineExecDescribeScriptOfScriptObject(*t_script_object_ref, t_object_ref -> getobject() == this);
```
Comment (livecodeali) [inline engine/src/exec-interface-object.cpp]: If `t_success` is false at some point here I guess we want to just exit the loop?

File: engine/src/exec-interface-object.cpp [hunk 3]
```diff
@@ -4704,3 +4707,126 @@ void MCObject::GetRevAvailableVariables(MCExecContext& ctxt, MCNameRef p_key, MC
...
+        t_success = MCEngineScriptObjectCreate(this, 0, &t_object_ref);
+    
+    if (t_success)
+        r_status = MCEngineExecDescribeScriptOfScriptObject(*t_object_ref);
+    
+    if (t_success &&
+        !MCExtensionConvertToScriptType(ctxt, r_status))
+    {
+        t_success = false;
+    }
```
Comment (livecodeali) [inline engine/src/exec-interface-object.cpp]: If `t_success` ends up false should this do `ctxt . Throw()` or something?

File: engine/src/exec-interface-object.cpp [hunk 4]
```diff
@@ -4704,3 +4707,126 @@ void MCObject::GetRevAvailableVariables(MCExecContext& ctxt, MCNameRef p_key, MC
...
+    
+    MCObjectListFree(t_object_list);
+    
+    r_descriptions = t_descriptions.Take();
+    
+    if (t_success &&
+        !MCExtensionConvertToScriptType(ctxt, r_descriptions))
+    {
+        t_success = false;
+    }
```
Comment (livecodeali) [inline engine/src/exec-interface-object.cpp]: And here? (if t_success ends up false should this do ctxt . Throw() or something?)

File: engine/src/exec-interface-object.cpp (hunk 3 above)
Comment (livecodeali) [reply engine/src/exec-interface-object.cpp]: In fact could the success variable be removed from this completely?

File: engine/src/exec-interface-object.cpp [hunk 5]
```diff
@@ -4657,7 +4660,7 @@ void MCObject::GetRevAvailableVariables(MCExecContext& ctxt, MCNameRef p_key, MC
     }
     
     // The handler code must be the first char of the string
-    const char_t t_handler_code = MCStringGetNativeCharAtIndex(t_key, 0);
```
Comment (montegoulding) [inline engine/src/exec-interface-object.cpp]: woops... forgot to revert this change. Will do

File: engine/src/exec-interface-object.cpp (hunk 3 above)
Comment (montegoulding) [reply engine/src/exec-interface-object.cpp]: lol... yes considering it's not being used at all

File: engine/src/exec-interface-object.cpp [hunk 6]
```diff
@@ -4704,3 +4707,126 @@ void MCObject::GetRevAvailableVariables(MCExecContext& ctxt, MCNameRef p_key, MC
...
+    t_success = MCArrayCreateMutable(&t_descriptions);
+    
+    MCObjectList *t_object_list;
+    t_object_list = nil;
+    
+    t_success = MCObjectListAppendObjectList(t_object_list, MCfrontscripts);
+    
+    for (MCObject *t_object = this; t_success && t_object != NULL; t_object = t_object -> parent)
+    {
+        t_success = MCObjectListAppendObjectAndBehaviors(t_object_list, t_object);
```
Comment (montegoulding) [inline engine/src/exec-interface-object.cpp]: @livecodeali now that I look at this again (was copied from revAvailableHandlers) I think two things:

1. I should make a utility function to get the object list and use it in both.
2. `t_success` here will be clobbered by the next iteration so it really should break if false each time

Comment (montegoulding) [reply engine/src/exec-interface-object.cpp]: actually I didn't notice the for break condition includes `t_success` so 2 isn't an issue. Will still make a utility function though.

**Comment by livecodeali** [comment]
@livecode-vulcan review ok ffaeb87

**Comment by livecode-vulcan** [comment]
💙 review by @livecodeali ok ffaeb87aa3655dad0c9987d1cdec8f9ab375b24c

**Comment by livecode-vulcan** [comment]
😎 test success ffaeb87aa3655dad0c9987d1cdec8f9ab375b24c

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/559)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/574)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/548)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/491)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/494)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/554)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/572)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/558)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/574)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/557)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/567)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/589)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/553)
